akka-stream/src/main/scala/akka/stream/impl/Timers.scala

```diff
@@ -269,5 +270,57 @@ import scala.concurrent.duration.{ Duration, FiniteDuration }
 
   }
 
+  final case class KeepAliveFromBuffer[T](size: Int, interval: FiniteDuration, extrapolate: T ⇒ Seq[T])
+    extends GraphStage[FlowShape[T, T]] {
+
+    require(size > 0, "The buffer size must be greater than 0.")
+
+    val in = Inlet[T]("KeepAliveFromBuffer.in")
+    val out = Outlet[T]("KeepAliveFromBuffer.out")
+
+    override val shape = FlowShape(in, out)
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+
+        private val buffer = new java.util.ArrayDeque[T](size)
+
+        override def preStart(): Unit = {
+          schedulePeriodically(None, interval)
+          pull(in)
+        }
+
+        override def onPush(): Unit = {
+          val elem = grab(in)
+          if (buffer.size() < size) buffer.addAll(extrapolate(elem).asJava)
+          else buffer.addLast(elem)
+
+          if (isAvailable(out)) push(out, buffer.removeFirst())
```

Comment:
is the throwing of an `NoSuchElementException` really expected here if the extrapolate emitted Nil?